### ISSUE TITLE
Errors should not be shared between queries

### ISSUE THREAD

**Issue by tobias-tengler** [facebook/relay#4637]
We're currently experimenting with Relay's new (experimental) approach to handling queries that contain errors. We've enabled both the `ENABLE_FIELD_ERROR_HANDLING` and `ENABLE_FIELD_ERROR_HANDLING_THROW_BY_DEFAULT` feature flag.

Overall it works really nicely and error-handling-clients are definitely the future. There is one major caveat though that still needs to be addressed in our opinion: Errors being shared between queries / fragments. 

> Note: I'm well aware that this is all experimental at the moment, but I still wanted to provide some early feedback from outside of Meta.

## What's the issue?

Let's imagine we have a page-level query like this:

```tsx
usePreloadedQuery(graphql`
  query PageQuery($id: ID!) {
    productById(id: $id) {
      ...BuyBox
    }
  }`,
  queryRef
);
```

and we also have another query that is only invoked, once you scroll down the page:

```tsx
useLazyLoadQuery(graphql`
  query LazyPageQuery($id: ID!) {
    productById(id: $id) {
      ...Reviews
    }
  }`,
 { id }
);
```

Currently, if the `PageQuery` is able to fetch the `productById` without an error, but the `LazyPageQuery` contains an error for the `productById` field, both queries will throw an error, since they both read out the errornous `productById` field.

This is pretty bad, because a temporary failure in the `LazyPageQuery` could take down other, previously successful, queries like the `PageQuery`.

### Reproduction

https://github.com/tobias-tengler/relay-errorhandling-error

## Expected behavior

I would expect that errors aren't shared between queries and fragments. In the above example I would expect only the `useLazyLoadQuery` hook to throw, since the error is associated with the `productById` field, which is read out by that hook. The `usePreloadedQuery` also reads out the `productById` field, but its query was able to resolve the `productById`, so it doesn't make sense to also propagate the error there and potentially kill an entire page, if the query itself completed without errors before.
In the same vain, fields that are associated with errors shouldn't update the store. Errornous fields are nulled and just how you would not want to share the error between queries, you also wouldn't want the `productById` field to suddenly become null in other places that were able to read it out without an error.

**Comment by captbaritone** [comment]
Thanks for providing the feedback! We've been thinking about this challenge internally as well. Today, every fragment or query rendered by relay render directly from the normalized store. That makes it difficult to show data in different states in different places. In fact, data consistency is a key priority of Relay, so generally this operates as a feature. 

That said, we agree that in many cases, it would be preferable to see the stale value than to blow up.

One idea we've considered is to avoid publishing errored fields to the store at all if a non-error field already exists. However this might be misleading in the primary query if you were expecting a change.

Cc @Itamar k

**Comment by itamark** [comment]
@tobias-tengler thanks so much for the feedback here. Even though this is experimental, it's really helpful to get this early feedback.

@captbaritone would storing stale value on errored fields (and providing them explicitly alongside the error state as, say, staleValue so engineers can fall back if they want to) - be anathema to Relay's the data consistency tenet?

**Comment by tobias-tengler** [comment]
I don't think errors fit the Relay's data consistency tenet. I want actual data to be synchronized across surfaces, but I find it very hard to find use cases where one would like an error in a different query to affect successful queries that came before it. If you want to have this behavior, you can implement it in user-space. It should definitely not be the default though. It's not only jarring for the developer, but also for the user. _Why did this thing in the sidebar suddenly have an error, if I clicked on this seemingly unrelated thing over here?_

Currently you use Relay under the assumption that something will only error if you re-execute it and an error occurs. Other queries being able to trigger an error for previously successful queries, breaks this assumption and is not really intuitive. 

@captbaritone My expectation would be that non-error fields in a response update the store, but errornous fields don't. 

@itamark The staleValue would also be strange for me. I don't want to have to handle actual values and stale values at every call-site, just because a query _somewhere_ could force my query into an error state.

**Comment by itamark** [comment]
This makes sense. Just checking - because error handling is meant to introduce correctness as well. In a case where the error is legitimate and an expected error occurs, is the expectation that the lazily loaded details page will get that data, but the previously successful query will still show the non-error-handled state? For example if a bestfriend deletes their profile or a product becomes unavailable.

As @captbaritone said we're actively discussing this behavior and this feedback is really helpful.

**Comment by itamark** [comment]
I was referring to my assumption that in a world with more explicit error handling, engineers would throw errors purposefully in order to handle them UI-side. Anyway - these are really good points. We're discussing this internally and I'll update - but in the meantime if you had any input on an alternative behavior, we'd love to hear it.

> My expectation would be that non-error fields in a response update the store, but errornous fields don't.

Just to confirm - this suggests that we still throw if an error happens in a particular query, but we wouldn't update the store with an error. Rather, we'd keep the previous value intact.

**Comment by tobias-tengler** [comment]
Yes. If a field is associated with an error, the new value (`null`) should not replace a previously successfully retrieved value in another query. Hooks attempting to read out a field associated with an error, that are "connected" to the query that produced the error, should throw.

**Comment by tobias-tengler** [comment]
After starting to roll-out semantic nullability, `@throwOnFieldError` and `@catch` internally, I feel like I have to revive this discussion again 😅

Syncing the error is terrible decision IMO!

A concrete case from our site: We have a shopping cart where you can change the quantity of a product in your cart. In the mutation response we're selecting some data necessary for tracking that's also selected on the actual page in a fragment.

If the server now returns an error for one of those fields in the mutation response this error will propagate to the fragment as well and cause the product in the cart to no longer render, since the fragment has thrown. And there's nothing I can do in the mutation to prevent the spreading of the error.

This is a horrible UX in line with resetting a form on failed form submission and working around this issue would introduce significant engineering overhead.

I really don't think it makes sense to erase existing data due an error, something unexpected, _somewhere_ in your application.
Sure, there are cases where you'd want this behavior, but it's a terrible default IMO.

**Comment by tobias-tengler** [comment]
I think maybe I first have to explain how we're treating errors. For us an untyped GraphQL error, as it comes in the `errors` array in the GraphQL response, represents an unexpected, potentially transient, developer-facing error. Something like a timeout, network issue, general authorization issue or straight up bug in the code. These errors are different from domain errors and error states. To stick with the example I gave, the mutation modifying the quantity failing because there is not enough product on hand, would return a typed error in the mutation field payload type. Similarly, validations or errors of the like that would be user-facing are modeled on the cart item schema type.

So there are domain errors, i.e. things we expect to go wrong or validations which can be considered user-facing data and which should be synced - no question. But GraphQL errors are fundamentally different in our view as they are something unexpected that happened in a particular operation.
If there is no data then we have to display the error to the user. But similarly to why you'd want to use `startTransition` in React, you wouldn't want to erase a perfectly fine UI, just because a different operation encountered such an unexpected error.

I'm interested if this is handled differently at Meta? Do you also model "domain errors" via GraphQL errors?

Propagating `null` values for error fields also increases the likelihood of running into missing data errors down the line, especially with transient errors. This is also something we're seeing and 

---

To your points:

1. _Some field errors should propagate to all readers. I think reasonable people can disagree if this should be the default, and ideally users would be able to set their own local default based on how they configure Relay._
  Agree
2. _Some field errors should be treated as errors for readers of the operation that observed the error, but other operations should be able to continue to read the stale value. (e.g. your tracking data example)_
  Agree
3. _If the store is sufficiently populated, some field errors should be allowed to be fully ignored in favor of stale data even for readers of the requesting operation._
  I'm not sure about this one. I think I don't agree. When executing an operation I virtually always want to see the current state. If there's an error in retrieving field X I don't want to show field X to the user. However, if I have retrieved and shown field X to the user I don't want to erase it due to an unrelated operation. In my view it should be up for the operation that introduces an error for an existing field to decide what should happen with that error. In the case of a mutation maybe I just want to show a toast to the user or set a state that grays out whatever I just interacted with, etc.
4. _The decision as to which fields can (or cannot) default to stale data is not one that the reader (fragment) or the writer (operation) can be asked to explicitly make for each field. Instead, this should be based on the importance/semantics of the fields. As such, this should be a configuration object/function set at the environment level which can decide which behavior is correct. I don't yet know what inputs such a configuration would want/need to be able to consider. Parent type? Field type? Field name? Operation type? Current stale value? I think we could provide all of those. Would that be sufficient?_
  Though one. I can see both the writer being in control and a global configuration. I need to think about it a bit more...
5. _The feature should not increase the likelihood of missing data errors. So, for example, a query with a field error should not be able to blindly fall back to stale data if the whole subtree below that field cannot be fulfilled from cache._ 
Absolutely. From what I've observed so far though, syncing the error is just as bad in this regard, especially if it's a transient one. It leads to some very strange bugs that you can almost not reproduce since you'd have to be in an error state and then follow a precise user journey to get to a point where the entity is refetched with a different selection set, causing the original location to now render with missing data.
